extra/tensors/tensors.factor

```diff
@@ -0,0 +1,257 @@
+! Copyright (C) 2019 HMC Clinic.
+! See http://factorcode.org/license.txt for BSD license.
+USING: accessors alien.c-types alien.data arrays
+concurrency.combinators grouping kernel locals math.functions
+math.ranges math multi-methods quotations sequences sequences.private
+specialized-arrays tensors.tensor-slice typed ;
+QUALIFIED-WITH: alien.c-types c
+SPECIALIZED-ARRAY: c:float
+IN: tensors
+
+! Tensor class definition
+TUPLE: tensor
+    { shape array }
+    { vec float-array } ;
+
+! Errors
+ERROR: shape-mismatch-error shape1 shape2 ;
+
+<PRIVATE
+
+! Construct a tensor of zeros
+: <tensor> ( shape seq -- tensor )
+    tensor boa ;
+
+: >float-array ( seq -- float-array )
+    [ ] float-array{ } map-as ;
+
+:: repetition ( shape const -- tensor )
```

Comment:
This could be a one-line definition if you wanted to:

```
: repetition ( shape const -- tensor )
    [ dup product ] dip <repetition> >float-array <tensor> ;
```